src/ai/ai_help_structs.cc

```diff
@@ -1100,6 +1101,82 @@ uint32_t PlayersStrengths::team_power(Widelands::TeamNumber tn) {
 	return team_powers_[tn];
 }
 
+// Make the decision whether own or player's team is better
+Widelands::DiplomacyAction PlayersStrengths::join_or_invite(const Widelands::PlayerNumber pl,
+                                                            const Time& gametime) {
+	if (this_player_number == pl || active_players_ < 3) {
+		return Widelands::DiplomacyAction::kRefuseJoin;
+	}
+
+	const Widelands::TeamNumber other_tn = all_stats_[pl].team_number;
+	if (this_player_team != 0 && this_player_team == other_tn) {
+		return Widelands::DiplomacyAction::kRefuseJoin;
+	}
+
+	int my_team_sc = 0;
+	int other_team_sc = 0;
+	bool me_alone = false;
+	bool other_alone = false;
+
+	if (other_tn == 0 || members_in_team(other_tn) == 1) {
+		other_team_sc = get_diplo_score(pl);
+		other_alone = true;
+	}
+	if (this_player_team == 0 || members_in_team(this_player_team) == 1) {
+		if (other_alone) {
+			return Widelands::DiplomacyAction::kInvite;
+		}
+		me_alone = true;
+	}
+
+	const bool can_invite = members_in_team(this_player_team) < active_players_ - 1;
+	const bool can_join = members_in_team(other_tn) < active_players_ - 1;
+	const Widelands::DiplomacyAction invite =
+	   can_invite ? Widelands::DiplomacyAction::kInvite : Widelands::DiplomacyAction::kRefuseJoin;
+	const Widelands::DiplomacyAction join =
+	   can_join ? Widelands::DiplomacyAction::kJoin : Widelands::DiplomacyAction::kRefuseJoin;
+
+	for (auto& item : all_stats_) {
+		if (item.first == pl || item.first == this_player_number || item.second.team_number == 0 ||
+		    item.second.defeated) {
+			continue;
+		}
+		if (can_invite && item.second.team_number == this_player_team) {
+			my_team_sc += item.second.players_diplomacy_score;
+		}
+		if (can_join && item.second.team_number == other_tn) {
+			other_team_sc += item.second.players_diplomacy_score;
+		}
+	}
+
+	if (!me_alone) {
+		my_team_sc /= members_in_team(this_player_team) - 1;
+	}
+
+	if (other_alone) {
+		if (other_team_sc - static_cast<int>(RNG::static_rand(10)) <= my_team_sc) {
+			verb_log_dbg_time(gametime,
+			                  "AI Diplomacy: Player(%d) with team score (%d) declines to invite "
+			                  "player (%d) with diploscore %d\n",
+			                  static_cast<unsigned int>(this_player_number), my_team_sc,
+			                  static_cast<unsigned int>(pl), other_team_sc);
+			return Widelands::DiplomacyAction::kRefuseJoin;
+		}
+	} else {
+		other_team_sc /= members_in_team(other_tn) - 1;
+	}
+
+	if (!can_invite || !can_join) {
```

Comment:
It is logged to have some trace if then the action is skipped because of the score. If both are allowed, then either choice will result in an action that is logged.

Reply:
I don't get it as far as I read the code nothing is skipped here but only logging takes place. the possible states of the 2 variables are either one of them is true or mostly both of them will be true they can't be both negative. but the condition is excluding the both true case.

Reply:
But this will be cleaned up then with the execution brought here.

Reply:
I understand the attempt but I think the logic might be wrong. case 0 0 can't happen but we would log. cases 1 0 or 0 1 we will log and some action happens. case 1 1 we won't log but action happens.  this seems to be inconsistent.